### ISSUE TITLE
Tools Tools Tools

### ISSUE THREAD

**Issue by cra0zy** [craftworkgames/MonoGame.Extended#47]
Simple question, should there be some tools included in MonoGame.Extended?

Example tools which could be useful are:
 - level editor
 - spritefont editor
 - xnb file previewer (to check if the file is good after compilation, would mostly be good for 3D Models)

**Comment by craftworkgames** [comment]
Excellent question. My ultimate goal is to help game developers make games. The slogan on [my blog](http://dylanwilson.net/) is "making tools for game developers".

My goal with MonoGame.Extended is to build the foundation to create other tools like the ones you've listed. It's not going to happen overnight, but it's definitely in the plan.

I'd be interested to know what kind of tools you'd value the most and what problems it could solve?

**Comment by cra0zy** [comment]
Well the 3 tools I listed above are the tools I would value the most, as for the problems they could solve:

1) Level Editor - besides being self explanatory, it would be very useful to include save and export, and export would work by setting a format that it would be able to export, ie:
```
<object id="%id" x="%x" y="%y"/>
```
and it auto exports it to that level format. Of course it would be more complex than that.

2) Spritefont editor - well I actually already created this: https://github.com/cra0zy/SFEditor I created it for easy font selection, easy setup of information and preview(which does not work with local fonts fully yet...)

3) Xnb file previewer - would be useful to check all the stuff that Pipeline Tool compiles to see if something needs modifying before adding it to your game... mostly useful for 3D Models because of Rotation check and Texture mapping

**Comment by craftworkgames** [comment]
@cra0zy These are some really good ideas.

> it would be very useful to include save and export, and export would work by setting a format

I've thought about this kind of thing before. I agree it would be a nice feature. It'd take some serious thinking to design it right.

> Spritefont editor - well I actually already created this: https://github.com/cra0zy/SFEditor 

That's cool. Editors like these are really useful. You should publish it on a blog or something. Put up some screenshots and let people know about it.

> Xnb file previewer - would be useful to check all the stuff that Pipeline Tool compiles to see if something needs modifying before adding it to your game... 

Yep. This would be good functionality to add to the Pipeline Tool itself. I've considered doing something like this at some point.

**Comment by cra0zy** [comment]
> That's cool. Editors like these are really useful. You should publish it on a blog or something. Put up some screenshots and let people know about it. Unfortunately, I don't have a Linux machine to try it out.

It's for Windows as well, I only couldn't compile it for Mac(don't have it)... also if you want I'll move it to this repo, I never liked that I created a whole repo for that tool.

**Comment by craftworkgames** [comment]
That looks sweet. I'll try build it tonight when I get home.

I'd rather keep tools out of this repo. I'm not sure how github intended things to be but in my experience having 1 repo per project is pretty common.

I'll fork your repo and take a look. If I make any changes I'll PR back. Looks like a great tool though. Nice job.

**Comment by craftworkgames** [comment]
I got it to compile and run out of Visual Studio 2015 on Windows 10. It was little fiddly because I forgot to do a submodule update and of course the Linux projects didn't compile. In the end though it was relatively easy. 

I'm too tired to do anything more significant tonight. It is cool though, I've never seen Xwt before. I'll have to remember it if I want to do a cross platform tool in the future.

**Comment by cra0zy** [comment]
Linux projects require gtk-sharp 2/3  to compile them.

gtk 2 sharp is available at: http://www.mono-project.com/download/#download-win
gtk 3 sharp that I used is from: https://github.com/openmedicus/gtk-sharp because they keep updated bindings for the newest version of GNOME, and they have to be installed by building from source

**Comment by craftworkgames** [comment]
You should really copy and paste those instructions into a "How to build" section of the README.md in the project. You took the time to write them for me, it'd be a shame to waste that effort.

**Comment by cra0zy** [comment]
What do you think on creating a repository named MonoGame.Extended.Tools that contains a collection of tools useful for MonoGame (would be pointing to tools with Git Submodules).

It could also have some kind of packager / installer for all the tools so that they don't have to be installed one by one.

**Comment by craftworkgames** [comment]
It's an interesting idea... maybe. What tools would you put in it initially?

I'm not sure how much value it adds over a simple list of tools with download links though. Not all tools are going to be open source and we'd have to maintain the repository to keep it up to date. 

Maybe something like [Chocolatey](https://chocolatey.org/) could be used to manage installing a group of tools. Most people probably don't always want all tools at once though as the list gets bigger.

**Comment by craftworkgames** [comment]
I'm off to bed now (really late here) so I'll take a look at your PR's tomorrow.

**Comment by craftworkgames** [comment]
I made a start on writing a tool with [Astrid Animator](http://dylanwilson.net/introducing-astrid-animator).

I've also got a [list of other tools on my blog](http://dylanwilson.net/list-of-game-development-content-creation-tools) that I wrote a while back.

Of course there's lots more we could do in this area, but one thing at a time.

I'm closing this issue because it's a little too general and has no clearly defined end.
